core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,141 @@
+                    instanceMetaDataService.getAllAvailableInstanceMetadataViewsByStackId(stack.getId()),
+                    request.getHostGroupName(),
+                    collectStartedInstancesFromCloudProvider(connector, ac, request.getAllInstancesInHostGroup()).stream()
+                            .map(CloudInstance::getInstanceId)
+                            .collect(Collectors.toSet()));
+
+
+            ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+
+            DetailedHostStatuses detailedHostStatuses = clusterHealthService.getDetailedHostStatuses();
+
+            List<InstanceMetadataView> startedInstanceMetadataWithServicesNotRunning = collectRecoveryCandidates(startedInstancesMetadata, detailedHostStatuses);
+
+            List<CloudInstance> instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(
+                    startedInstanceMetadataWithServicesNotRunning, stack.getStack());
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(stack.getId(),
+                    instancesWithServicesNotRunning, request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect downscale recovery candidates for stack: %s. " +
+                    "Proceeding with subsequent flow.", request.getResourceId());
+            LOGGER.error(errorMessage, e);
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector<?> connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses) {
+        // TODO Predicate will change for upscale and downscale, once MM is enabled before downscale decommission
+        Predicate<InstanceMetadataView> instanceHealthCheck =  im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            return detailedHostStatuses.isHostInMaintenanceMode(hostName)
+                || detailedHostStatuses.isHostDecommissioned(hostName)
```

Comment:
Should isHostDecommissioned be checking for DECOMMISSIONED instead of !COMMISSIONED?

Reply:
Will add more comments after looking in more detail. I don't think this maps to what was outlined in the writeup. (More in the upscale candidate collection comment)

Reply:
Updated to match the items discussed in the wrtie-up